crates/swc_plugin_runner/src/cache.rs

```diff
@@ -0,0 +1,153 @@
+    let mut root_path = if let Some(root) = filesystem_cache_root {
+        Some(PathBuf::from(root))
+    } else if let Ok(cwd) = current_dir() {
+        Some(cwd.join(".swc"))
+    } else {
+        None
+    };
+
+    if let Some(root_path) = &mut root_path {
+        root_path.push("plugins");
+        root_path.push(MODULE_SERIALIZATION_VERSION);
+
+        return FileSystemCache::new(&root_path).ok();
+    }
+
+    None
+}
+
+/// Create a new cache instance if not intialized. This can be called multiple
+/// time, but any subsequent call will be ignored.
+///
+/// This fn have a side effect to create path to cache if given path is not
+/// resolvable. If root is not specified, it'll generate default root for
+/// cache location.
+///
+/// If cache failed to initialize filesystem cache for given location
+/// it'll be serve in-memory cache only.
+pub fn init_plugin_module_cache_once(filesystem_cache_root: &Option<String>) {
+    PLUGIN_MODULE_CACHE.inner.get_or_init(|| {
+        Mutex::new(CacheInner {
+            fs_cache: create_filesystem_cache(filesystem_cache_root),
+            memory_cache: Default::default(),
+        })
+    });
+}
+
+impl PluginModuleCache {
+    /// DO NOT USE unless absolutely necessary. This is mainly for testing
+    /// purpose.
+    pub fn new() -> Self {
+        PluginModuleCache {
+            inner: OnceCell::from(Mutex::new(Default::default())),
+            instantiation_lock: Mutex::new(()),
+        }
+    }
+
+    /// Load a compiled plugin Module from speficied path.
+    /// Since plugin will be initialized per-file transform, this function tries
+    /// to avoid reading filesystem per each initialization via naive
+    /// in-memory map which stores raw bytecodes from file. Unlike compiled
+    /// bytecode cache for the wasm, this is volatile.
+    ///
+    /// ### Notes
+    /// [This code](https://github.com/swc-project/swc/blob/fc4c6708f24cda39640fbbfe56123f2f6eeb2474/crates/swc/src/plugin.rs#L19-L44)
+    /// includes previous incorrect attempt to workaround file read issues.
+    /// In actual transform, `plugins` is also being called per each transform.
+    pub fn load_module(&self, binary_path: &Path) -> Result<Module, Error> {
+        let binary_path = binary_path.to_path_buf();
+        let mut inner_cache = self.inner.get().expect("Cache should be available").lock();
+
+        // if constructed Module is available in-memory, directly return it.
+        // Note we do not invalidate in-memory cache currently: if wasm binary is
+        // replaced in-process lifecycle (i.e devserver) it won't be reflected.
+        let in_memory_module = inner_cache.memory_cache.modules.get(&binary_path);
+        if let Some(module) = in_memory_module {
+            return Ok(module.clone());
+        }
+
+        let module_bytes =
+            std::fs::read(&binary_path).context("Cannot read plugin from specified path")?;
```

Comment:
Unlike previous implementation, this doesn't try to locate in-memory bytecode cache anymore: since we expect in-memory cache should be available once after we read this file from fs / or fscache for compiled bytecode.